Parser/myreadline.c

```diff
@@ -392,9 +392,11 @@ PyOS_Readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
         }
     }
 
-    _PyOS_ReadlineTState = tstate;
-    Py_BEGIN_ALLOW_THREADS
+
     PyThread_acquire_lock(_PyOS_ReadlineLock, 1);
+    Py_BEGIN_ALLOW_THREADS
+    _PyOS_ReadlineTState = tstate;
```

Comment:
Maybe set _PyOS_ReadlineTState while you are still holding the GIL, before Py_BEGIN_ALLOW_THREADS. It should not make a big difference, since other threads should be blocked at PyThread_acquire_lock(_PyOS_ReadlineLock, 1) :-)

Reply:
Alrighty :-)

Reply:
So we have to release the lock outside the GIL and preferably take the lock too, otherwise it [deadlocks](https://github.com/python/cpython/pull/123323#discussion_r1739368059).

It fails on the free-threaded as according to @colesbury the taking of the lock isn't safe outside the GIL.

I'm a little confused as theoretically if it isn't safe it should fail on the normal build too but I'm not entirely sure how that works.